File: Time Measurement/Go/execute.go
package main

import "fmt"
import "os"
import "os/exec"
import "time"
import "sync"
import "math"

type execute struct{
	d *data
	o *os.File
	e *os.File
	r *os.File
	res string
	ec int
	opened map[string]*os.File
};



func ex(d *data) {
	x:=execute{
		d:d,
		ec:0,
		opened:map[string]*os.File{},
	}
	x.o=x.co2f(d.out,os.Stdout)
	x.e=x.co2f(d.err,os.Stderr)
	x.r=x.ro2f(d.result)

	switch d.multiple {
		case mmNone:   x.single()
		case mmSerial: x.serial()
		case mmSpawn:  x.spawn()
		case mmThread: x.thread()
	}

	fmt.Fprint(x.r,x.res)
	x.o.Close()
	x.e.Close()
	x.r.Close()

	if x.ec==-1 { os.Exit(1) } else { os.Exit(x.ec) }
}

func (x *execute) single() {
	p:=x.sp(x.d.command)

	st:=time.Now()
	p.run()
	en:=time.Now()

	x.res=clean(fmt.Sprint(`
		time: `,x.descTime(st,en),`
		process id: `,p.pid,`
		`,p.descEC(),`
	`))
	x.ec=p.ec
}

func (x *execute) serial() {
	pl:=x.spMultiple(x.d.command)
	var lp *sp=pl[len(pl)-1]

	st:=time.Now()
	for _,p:=range pl {
		p.run()
		if p.ec!=0 {
			lp=p
			break
		}
	}
	en:=time.Now()

	x.res=fmt.Sprintf("time: %s\n",x.descTime(st,en))
	for _,p:=range pl {
		pid:=fmt.Sprintf("%d",p.pid)
		if p.pid<0 { pid="N/A" }
		x.res+=fmt.Sprintf("process%d id: %s\n",p.order,pid)
	}
	x.res+=lp.descEC()+"\n"

	x.ec=lp.ec
}

func (x *execute) spawn() {
	pl:=x.spMultiple(x.d.command)

	st:=time.Now()
	for _,p:=range pl { p.start() }
	for _,p:=range pl { p.wait() }
	en:=time.Now()

	x.collect(pl,st,en)
}

func (x *execute) thread() {
	pl:=x.spMultiple(x.d.command)
	wg:=&sync.WaitGroup{}

	st:=time.Now()
	for _,p:=range pl {
		wg.Add(1)
		go x.threadFunc(p,wg)
	}
	wg.Wait()
	en:=time.Now()

	x.collect(pl,st,en)
}

func (x *execute) threadFunc(p *sp,wg *sync.WaitGroup) {
	p.run()
	wg.Done()
}

type sp struct {
	cmd *exec.Cmd
	parent *execute
	order int
	pid int
	ec int
}
func (x *execute) sp(args []string) *sp {

	var cmd *exec.Cmd
	if len(args)==1 { cmd=exec.Command(args[0]) } else
	{ cmd=exec.Command(args[0],args[1:]...) }
	cmd.Stdin=os.Stdin
	if x.o!=nil { cmd.Stdout=x.o }
	if x.e!=nil { cmd.Stderr=x.e }

	return &sp{
		cmd:cmd,
		parent:x,
		order:0,
		pid:-1,
		ec:0,
	}
}
func (x *execute) spMultiple(commands []string) []*sp {
	sh,has:=os.LookupEnv("SHELL")
	if !has {sh="sh"}
	l:=make([]*sp,len(commands))
	for n,c:=range commands {
		var p=x.sp([]string{sh,"-c",c})
		p.order=n+1
		l[n]=p
	}
	return l
}
func (x *execute) collect(pl []*sp,st time.Time,en time.Time) {
	x.res=fmt.Sprintf("time: %s\n",x.descTime(st,en))
	for _,p:=range pl {
		x.res+=fmt.Sprintf("process%d id: %d\n",p.order,p.pid)
		x.res+=p.descEC()+"\n"
		if p.ec>x.ec { x.ec=p.ec }
	}
}

func (p *sp) start() {
	e:=p.cmd.Start()
	if e!=nil { error("起動に失敗しました") }
	p.pid=p.cmd.Process.Pid
}
func (p *sp) wait() {
	p.cmd.Wait()
	p.ec=p.cmd.ProcessState.ExitCode()
}
func (p *sp) run() {
	p.start()
	p.wait()
}
func (p *sp) descEC() string {
	var s string
	if p.ec==-1 { s="terminated due to signal" } else
	{ s=fmt.Sprint("exit code: ",p.ec) }
	return s
}



func (x *execute) co2f(d string,inherit *os.File) *os.File {
	switch d {
		case "inherit": return inherit
		case "discard": return nil
		default: return nil
	}
}
func (x *execute) ro2f(d string) *os.File {
	switch d {
		case "stdout": return os.Stdout
		case "stderr": return os.Stderr
		default: return x.fh(d)
	}
}
func (x *execute) fh(path string) *os.File {
	ef,has := x.opened[path]
	if has { return ef }
	f,e := os.OpenFile(path,os.O_APPEND|os.O_CREATE|os.O_WRONLY,0644)
	if e != nil { error("指定したパスには書き込みできません: "+path) }
	x.opened[path]=f
	return f
}



func (x execute) descTime(st time.Time,en time.Time) string {
	t:=""
	r:=float64(en.Sub(st))/(3600*1e+9)
	v:=math.Floor(r)
	if v>=1 { t+=fmt.Sprintf("%.0fh ",v) }
	r=(r-v)*60
	v=math.Floor(r)
	if v>=1 { t+=fmt.Sprintf("%.0fm ",v) }
	r=(r-v)*60
	v=math.Floor(r)
	if v>=1 { t+=fmt.Sprintf("%.0fs ",v) }
	r=(r-v)*1000
	t+=fmt.Sprintf("%07.3fms",r)
	return t
}

File: Time Measurement/JavaScript/execute.js
const
cp=require("child_process"),
fs=require("fs"),
exit=require("process").exit,
now=require("perf_hooks").performance.now,
EOL=require("os").EOL,

{MMNone,MMSerial,MMSpawn,data,clean}=require("./lib.js");

exports.main=async (rd=new data())=>{
	d=rd;
	o=co2f(d.out);
	e=co2f(d.err);
	r=ro2f(d.result);

	switch (d.multiple) {
		case MMNone:   single();      break;
		case MMSerial: serial();      break;
		case MMSpawn:  await spawn(); break;
	}

	fs.writeSync(r,res);
	fs.closeSync(r);
	exit(ec??1);
}

var d,o,e,r;
var ec=0,res="";

function single() {
	let p=new SP(d.command);

	let st=now();
	p.run();
	let en=now();

	res=clean(`
		time: ${descTime(en-st)}
		process id: ${p.pid}
		${p.descEC()}
	`);
	ec=p.ec;
}

function serial() {
	let pl=SP.multiple(d.command);
	var lp=pl[pl.length-1];

	let st=now();
	for (p of pl) {
		p.run();
		if (p.ec!=0) {
			lp=p;
			break;
		}
	}
	let en=now();

	res=[
		"time: "+descTime(en-st),
		...pl.map(p=>`process${p.order} id: ${p.pid<0?"N/A":p.pid}`),
		p.descEC(),""
	].join(EOL);
	ec=lp.ec;
}

async function spawn() {
	let pl=SP.multiple(d.command);

	let st=now();
	await Promise.all(pl.map(p=>p.start()));
	let en=now();

	res=[
		`time: ${descTime(en-st)}`,
		...pl.flatMap(p=>{
			if (p.ec>ec) ec=p.ec;
			return [`process${p.order} id: ${p.pid}`,p.descEC()];
		}),""
	].flat().join("\n");
}

class SP {
	args=[];

	order=0;
	pid=-1;
	ec=0;
	signal=null;
	descEC() {
		return this.ec==null?"terminated due to "+(this.signal??"signal"):"exit code: "+this.ec;
	}

	constructor(args) {
		if (typeof(args)=="string") this.args=[args,{stdio:["inherit",o,e],shell:true}];
		else {
			let f=args.shift();
			this.args=[f,args,{stdio:["inherit",o,e]}];
		}
	}
	static multiple(commands) {
		var n=1,c;
		let l=[];
		for (c of commands) {
			let p=new SP(c);
			p.order=n;
			l.push(p);
			n++;
		}
		return l;
	}
	start() {
		let t=this;
		return new Promise(resolve=>{
			let p=cp.spawn(...t.args);
			t.pid=p.pid;
			p.on("exit",(ec,signal)=>{
				t.ec=ec;
				t.signal=signal;
				resolve();
			});
		});
	}
	run() {
		let s=cp.spawnSync(...this.args);
		this.pid=s.pid;
		this.ec=s.status;
		this.signal=s.signal;
	}
}



function co2f(d) {
	switch (d) {
		case "inherit": return "inherit";
		case "discard": return "ignore";
		default:        return fh(d);
	}
}

function ro2f(d) {
	switch (d) {
		case "stdout": return 1;
		case "stderr": return 2;
		default:       return fh(d);
	}
}

let opened={};
function fh(path) {
	if (opened[path]) return opened[path];
	try{
		let io=fs.openSync(path,"a");
		opened[path]=io;
		return io;
	}
	catch(e) { error("指定したパスには書き込みできません: "+path); }
}



function descTime(msec) {
	var t="";
	var r=msec/(3600*1e+3),v=Math.floor(r);
	if (v>=1) t+=`${v}h `;
	r=(r-v)*60,v=Math.floor(r);
	if (v>=1) t+=`${v}m `;
	r=(r-v)*60,v=Math.floor(r);
	if (v>=1) t+=`${v}s `;
	r=(r-v)*1000;
	let ms=r.toFixed(3);
	t+=`${"0".repeat(7-ms.length)}${ms}ms`;
	return t;
}